packages/oauth/src/index.ts

```diff
@@ -0,0 +1,555 @@
+  // failure is given control when handleCallback() fails at any point.
+  // when provided, this function must complete the callbackRes.
+  // default:
+  // serve a generic "Error" web page (show detailed cause in development)
+  failure?: (
+    error: CodedError,
+    options: InstallURLOptions,
+    callbackReq: IncomingMessage,
+    callbackRes: ServerResponse,
+  ) => void;
+}
+
+// Response shape from oauth.v2.access - https://api.slack.com/methods/oauth.v2.access#response
+interface OAuthV2Response {
+  ok: boolean;
+  app_id: string;
+  authed_user: {
+    id: string,
+    scope?: string,
+    access_token?: string,
+    token_type?: string,
+  };
+  scope: string;
+  token_type: string;
+  access_token: string;
+  bot_user_id: string;
+  team: { id: string, name: string };
+  enterprise: { name: string, id: string } | null;
+  error?: string;
+  incoming_webhook?: {
+    url: string,
+    channel: string,
+    channel_id: string,
+    configuration_url: string,
+  };
+  response_metadata: object;
+}
+
+// Response shape from oauth.access - https://api.slack.com/methods/oauth.access#response
+interface OAuthV1Response {
+  ok: boolean;
+  access_token: string;
+  // scope parameter isn't returned in workspace apps
+  scope: string;
+  team_name: string;
+  team_id: string;
+  enterprise_id: string | null;
+  // if they request bot user token
+  bot?: { bot_user_id: string, bot_access_token: string };
+  incoming_webhook?: {
+    url: string,
+    channel: string,
+    channel_id: string,
+    configuration_url: string,
+  };
+  response_metadata: object;
+  error?: string;
+  // app_id is currently undefinined but leaving it in here incase the v1 method adds it
+  app_id: string | undefined;
+}
+
+export interface StateStore {
+  // Returned Promise resolves for a string which can be used as an
+  // OAuth state param.
+  generateStateParam: (now: Date, metadata?: string) => Promise<string>;
+
+  // Returned Promise resolves for metadata that was stored in the state
+  // param, or undefined when there was no metadata. The Promise rejects
+  // with a CodedError when the state is invalid.
+  verifyStateParam: (now: Date, state: string) => Promise<string | undefined>;
+}
+
+// State object structure
+interface StateObj {
+  now: Date;
+  metadata?: string;
+}
+
+// default implementation of StateStore
+class ClearStateStore implements StateStore {
+  private stateSecret: string;
+  public constructor(stateSecret: string | undefined) {
+    if (stateSecret == null) {
+      throw new Error('You must provide a State Secret to use the built-in state store');
+    }
+    this.stateSecret = stateSecret;
+  }
+
+  // TODO: Question, why do the params below need to have types definied
+  // instead of getting those types from StateStore interface
+  public generateStateParam(now: Date, metadata?: string): Promise<string> {
+    const state = sign({ metadata, now: now.toJSON() }, this.stateSecret);
+    return new Promise<string>((resolve) => {
```

Comment:
make the function `async` and just return the value (no need to construct a Promise).